### PR TITLE
ccda author and dates

### PR DESCRIPTION
**Pull request by sjpadgett**
- 3 chances for author
- datetime validation changes

### CODE REVIEW

**Comment by sjpadgett** [comment]
@adunsulag Sorry, fell asleep!:)
Dates still feels sloppy but still working.
Author is correct for document but still needs component authors looked into

**Comment by adunsulag** [comment]
@sjpadgett Thanks Jerry!  I briefly looked over the code and will take a deeper dive tomorrow morning.

File: interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/EncounterccdadispatchTable.php [hunk 1]
```diff
@@ -2224,18 +2234,23 @@ public function getCarecoordinationModuleSettingValue($field_name)
...
+        JOIN module_configuration AS conf ON conf.field_value=f.id AND mo.mod_id=conf.module_id
+        WHERE conf.field_name=?";
+        } elseif (is_string($field_name)) {
+            $query = "SELECT u.title, u.fname, u.mname, u.lname, u.npi, u.street, u.city, u.state, u.zip, CONCAT_WS(' ','',u.phonew1) AS phonew1, u.organization, u.specialty, conf.field_name, mo.mod_name, lo.title as  physician_type, SUBSTRING(lo.codes, LENGTH('SNOMED-CT:')+1, LENGTH(lo.codes)) as  physician_type_code, uuid
+        FROM users AS u
+        LEFT JOIN list_options AS lo ON lo.list_id = 'physician_type' AND lo.option_id = u.physician_type
+        JOIN modules AS mo ON mo.mod_directory='Carecoordination'
+        JOIN module_configuration AS conf ON conf.field_value=u.id AND mo.mod_id=conf.module_id
+        WHERE conf.field_name=?";
+        } elseif (is_int($field_name)) {
```
Comment (adunsulag) [inline interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/EncounterccdadispatchTable.php:2249]: If I'm following this correctly, you treat the field_name as the user id if its an integer?

**Comment by sjpadgett** [comment]
okay just pushed your +1 in author fix.

File: interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/CcdaServiceRequestModelGenerator.php [hunk 2]
```diff
@@ -84,11 +84,11 @@ public function create_data($pid, $encounter, $sections, $components, $recipient
             <telecom use='WP' value='" . htmlspecialchars($assignedEntity['telecom'] ?? '', ENT_QUOTES) . "'/>
             <representedOrganization>
                 <name>" . htmlspecialchars($representedOrganization['name'] ?? '', ENT_QUOTES) . "</name>
-                <telecom use='WP' value='" . htmlspecialchars($representedOrganization['telecom'] ?? '', ENT_QUOTES) . "'/>
-                <streetAddressLine>" . htmlspecialchars($representedOrganization['streetAddressLine'] ?? '', ENT_QUOTES) . "</streetAddressLine>
+                <telecom use='WP' value='" . htmlspecialchars($representedOrganization['phone'] ?? '', ENT_QUOTES) . "'/>
```
Comment (adunsulag) [inline interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/CcdaServiceRequestModelGenerator.php:87]: I'm guessing this fixes the phone issues we were running into?

Comment (sjpadgett) [reply interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/CcdaServiceRequestModelGenerator.php:87]: yep, you were looking at translate but this never made it there. I was using the wrong telephone template in header author.

Comment (adunsulag) [reply interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/CcdaServiceRequestModelGenerator.php:87]: awesome!  I'm glad you found it!

File: interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/EncounterccdadispatchTable.php (hunk 1 above)
Comment (sjpadgett) [reply interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/EncounterccdadispatchTable.php:2249]: yep

**Comment by sjpadgett** [comment]
I'm going to fix the referral then hit bed. Will pickup on dates tomorrow

**Comment by sjpadgett** [comment]
@adunsulag where are you seeing the provider req for referral?

**Comment by adunsulag** [comment]
@sjpadgett 
SLI Test Script
image

image

**Comment by sjpadgett** [comment]
okay give me a quick call

**Comment by adunsulag** [comment]
@sjpadgett There were a couple of things I had to fix.  I put out a pull request to your repository with the fixes here: https://github.com/sjpadgett/openemr/pull/21

I'm passing Alice with the changes.  Great job on all of the timezone changes!  The scenario MDHT fails when I don't override the time zone because its using the current timezone settigs.  The ETT validator even caught that the timezone was not matching the scenario, when I switched to the 06/22/2015 11:00 AM EST everything passed.   Very pleased with how that turned out so great job.

I gotta run for an hour or so but I'll then go through and verify Jeremy and Happy Kid to make sure that's working on the ETT, then I'll do the visual inspection on the timezones as well to make sure nothing's missing. Looks like phone numbers are working as well (from what I quickly saw).

**Comment by sjpadgett** [comment]
added primary diagnosis code to documentationOf
```
<serviceEvent classCode="PCPR">
      <code code="386661006" displayName="Fever Finding" codeSystem="2.16.840.1.113883.6.96" codeSystemName="SNOMED CT"/>
      
```

**Comment by sjpadgett** [comment]
add time zone to encounter date
```
<effectiveTime value="202006221000-0400"/>
```

**Comment by sjpadgett** [comment]
working on componentOf
```
"primary_diagnosis": {
        "encounter_date": "2020-06-22 10:00:00-0400",
        "encounter_end_date": "2020-06-22 10:30:00-0400",
        "code": "386661006",
        "code_type": "SNOMED CT",
        "text": "Fever (finding)",
        "status": "Completed"
    }
```

**Comment by sjpadgett** [comment]
I am just so pleased with this since @adunsulag did his style improvements.

Summarization-of-Episode-Note

**Comment by sjpadgett** [comment]
@adunsulag 
Did a lot of work on dates, time zones, author etc. I added to translate time a utc -0400 offset for any date time precision tz, All b1 scenarios pass.
I changed the encounter and note author time to reflect the CCM author time. I believe this may be a mistake. We should discuss tomorrow.

Unsure if you want this merged before you test so, merge if you want or need and I'm not around in the a.m.

File: ccdaservice/oe-blue-button-generate/lib/translate.js [hunk 3]
```diff
@@ -75,12 +75,18 @@ var precisionToFormat = {
...
+    let result = '';
     var m = moment.parseZone(input.date);
     if (m._isValid !== true) {
         m = moment(input.date, "YYYYMMDD HH:mm:ss")
     }
     let formatSpec = precisionToFormat[input.precision];
-    let result = m.format(formatSpec);
+    if (input.precision === 'tz') {
+        formatSpec = precisionToFormat['tz'];
+        result =  m.utcOffset('-0400').format(formatSpec);
```
[the remark this answers is not in the document]
Comment (adunsulag) [reply ccdaservice/oe-blue-button-generate/lib/translate.js:86]: When I mean sending it for the entire document, I'm meaning sending it in php the timezone offset as part of the xml (-400) in this case.

[the remark this answers is not in the document]
Comment (sjpadgett) [reply ccdaservice/oe-blue-button-generate/lib/translate.js:86]: We'd have to bring in the moment-timezone library to set this global. Using environment var and the browsers doesn't seem to give the timezone as they once did. 
We could grab info from PHP and set but for now, this gets us there.

[the remark this answers is not in the document]
Comment (adunsulag) [reply ccdaservice/oe-blue-button-generate/lib/translate.js:86]: yep, I agree.  I'll do a smoke screen test run against this PR and if everything is looking good I'll bring it in for a full test run today.

**Comment by adunsulag** [comment]
@sjpadgett Just did a test run with b1 on the send and everything is passing the validator. Great Job!  I also went through all the XML and the timezone is populated for authors for everything but the lab results.  I'll merge this in and fix the lab results authorship piece so no biggy there. 

I thought when I had read the companion guide that we only want the encompassing encounter on Encounter Summary type documents such as the Referral form.  So we would skip it for CCD since it covers multiple encounters.  I'm basing that on page 37-38 of the Companion Guide: http://www.hl7.org/ccdasearch/pdfs/Companion_Guide.pdf

**Comment by sjpadgett** [comment]
I test for multiple encounters see L-2530 of generator `if (isOne(all.encounter_list.encounter) === 1)`. If is one, componentOf shows. Regardless of document type, if documents focus is a single encounter then a componentOf with the encompassingEncounter MAY be included.

**Comment by adunsulag** [comment]
ok, sounds good.  I'll remove my docType checks then in my PR.

**Pull request merged**